lib/SILOptimizer/Transforms/SILMem2Reg.cpp

```diff
@@ -1160,8 +1295,16 @@ void StackAllocationPromoter::endLexicalLifetime(BlockSetVector &phiBlocks) {
     case AvailableValuesKind::In: {
       if (auto *inst = deinitializationPoints[bb]) {
         auto values = getLiveInValues(phiBlocks, bb);
-        endLexicalLifetimeBeforeInst(asi, /*beforeInstruction=*/inst, ctx,
-                                     *values);
+        if (isa<EndBorrowInst>(inst)) {
+          if (!canEndLexicalLifetime(*values)) {
```

Comment:
How can there be a live-in value and a deinitialization point but we can't end the lifetime at it?  Might this be happening because `lastStoreInst` isn't set to `llvm::None` in `StackAllocationPromoter::promoteAllocationInBlock` when looking at an `EndBorrowInst` like happens when looking at `DestroyValueInst`s?

Reply:
Not all store_borrow to lexical alloc_stack will have a borrow to end. When the source `isGuaranteedLexicalValue` we wouldn't have inserted a lexical borrow that needs to be ended. I'll add a comment here.

Reply:
I mark lastStoreInst->isStorageValid = false when we see an end_borrow for the store_borrow, in such cases the store_borrow shouldn't be in the `intializationPoint`

Reply:
I'm seeing where `runningVals->isStorageValid` gets set to false: https://github.com/apple/swift/pull/59541/commits/8c9c61a956e4047826079627fa05fab90e7e8a33#diff-a0cdc552f97b0b66a7c8f63788292539566ab2a53affa110bc00f904df39ecf9R1833 , but not where `lastStoreInst->isStorageValid` does.  Am I missing it?

Reply:
Huh. I didn't realize we have 2 of those now. I'll take a look and either remove it if redundant or set it to false. Thanks

Reply:
Also see - https://github.com/apple/swift/pull/59541/commits/0c2fd6b419babfb4639eace7d060f5d096883f8f 
and https://github.com/apple/swift/pull/59541/commits/0c2fd6b419babfb4639eace7d060f5d096883f8f#r952018811. I am removing the special case for destroy_value.

Reply:
Thanks!  Setting it to false like in the handling of `DestroyAddrInst` sounds right.  They're not redundant.  One tracks the most recent store instruction.  One tracks which values are live.

Reply:
Got rid of the redundant flag `isStorageValid` on lastStoreInst in the latest update.